cirq-core/cirq/ops/kraus_channel.py

```diff
@@ -0,0 +1,105 @@
+from typing import Any, Dict, Iterable, Tuple, Union
+import numpy as np
+
+from cirq import linalg, protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+# TODO(#3241): support qudits and non-square operators.
+class KrausChannel(raw_types.Gate):
+    """A generic channel that can record the index of its selected operator.
+
+    Args:
+        kraus_ops: a list of Kraus operators, formatted as numpy array.
+            Currently, only square-matrix operators on qubits (not qudits) are
+            supported by this type.
+        key: an optional measurement key string for this channel. Simulations
+            which select a single Kraus operator to apply will store the index
+            of that operator in the measurement result list with this key.
+        validate: if True, validate that `kraus_ops` describe a valid channel.
+            This validation can be slow; prefer pre-validating if possible.
+    """
+
+    def __init__(
+        self,
+        kraus_ops: Iterable[np.ndarray],
+        key: Union[str, value.MeasurementKey, None] = None,
+        validate: bool = False,
+    ):
+        kraus_ops = list(kraus_ops)
+        if not kraus_ops:
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
```

Comment:
Should probably also check that `kraus_ops[0].shape[0] == kraus_ops[0].shape[1]`.

Reply:
Done. Also updated a few adjacent checks to use `shape` instead of `size`.